cpp/src/arrow/flight/server.cc

```diff
@@ -64,9 +69,11 @@ class FlightMessageReaderImpl : public FlightMessageReader {
  public:
   FlightMessageReaderImpl(const FlightDescriptor& descriptor,
                           std::shared_ptr<Schema> schema,
+                          std::unique_ptr<ipc::DictionaryMemo> dict_memo,
```

Comment:
See #4319 for a clean, dictionary-compatible, re-implementation of `FlightMessageReader` based on `ipc::MessageReader`.